File: src/main/java/aurocosh/divinefavor/common/custom_data/world/data/altars/AltarsDataSerializer.java
package aurocosh.divinefavor.common.custom_data.world.data.altars;

import aurocosh.divinefavor.common.lib.interfaces.INbtSerializer;
import aurocosh.divinefavor.common.registry.mappers.ModMappers;
import aurocosh.divinefavor.common.spirit.base.ModSpirit;
import aurocosh.divinefavor.common.util.UtilBlockPos;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.math.BlockPos;

import java.util.*;

// Handles the actual read/write of the nbt.
public class AltarsDataSerializer implements INbtSerializer<AltarsData> {
    private static String TAG_ALTAR_POSITIONS = "AltarPositions";

    @Override
    public void serialize(NBTTagCompound nbt, AltarsData instance) {
        NBTTagCompound altarNbt = new NBTTagCompound();

        Map<Integer, Set<BlockPos>> positions = instance.getAllPositions();
        for (Map.Entry<Integer, Set<BlockPos>> entry : positions.entrySet()) {
            int spiritId = entry.getKey();
            List<BlockPos> altarLocations = new ArrayList<>(entry.getValue());
            ModSpirit spirit = ModMappers.spirits.get(spiritId);
            int[] posArray = UtilBlockPos.serialize(altarLocations);
            String spiritName = spirit.getRegistryName().toString();
            altarNbt.setIntArray(spiritName, posArray);
        }
        nbt.setTag(TAG_ALTAR_POSITIONS, altarNbt);
    }

    @Override
    public void deserialize(NBTTagCompound nbt, AltarsData instance) {
        if(!nbt.hasKey(TAG_ALTAR_POSITIONS))
            return;
        NBTTagCompound altarNbt = nbt.getCompoundTag(TAG_ALTAR_POSITIONS);
        List<ModSpirit> spirits = ModMappers.spirits.getValues();
        Map<Integer, Set<BlockPos>> positions = new HashMap<>();
        for (ModSpirit spirit : spirits) {
            String spiritName = spirit.getRegistryName().toString();
            if(!altarNbt.hasKey(spiritName))
                return;
            int[] posArray = altarNbt.getIntArray(spiritName);
            List<BlockPos> altarLocations = UtilBlockPos.deserialize(posArray);
            positions.put(spirit.getId(), new HashSet<>(altarLocations));
        }
        instance.setAllPositions(positions);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/custom_data/world/data/altars/AltarsDataSerializer.kt
package aurocosh.divinefavor.common.custom_data.world.data.altars

import aurocosh.divinefavor.common.lib.interfaces.INbtSerializer
import aurocosh.divinefavor.common.registry.mappers.ModMappers
import aurocosh.divinefavor.common.spirit.base.ModSpirit
import aurocosh.divinefavor.common.util.UtilBlockPos
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.math.BlockPos

import java.util.*

// Handles the actual read/write of the nbt.
class AltarsDataSerializer : INbtSerializer<AltarsData> {

    override fun serialize(nbt: NBTTagCompound, instance: AltarsData) {
        val altarNbt = NBTTagCompound()

        val positions = instance.allPositions
        for ((spiritId, value) in positions) {
            val altarLocations = ArrayList(value)
            val spirit = ModMappers.spirits[spiritId]
            val posArray = UtilBlockPos.serialize(altarLocations)
            val spiritName = spirit.registryName!!.toString()
            altarNbt.setIntArray(spiritName, posArray)
        }
        nbt.setTag(TAG_ALTAR_POSITIONS, altarNbt)
    }

    override fun deserialize(nbt: NBTTagCompound, instance: AltarsData) {
        if (!nbt.hasKey(TAG_ALTAR_POSITIONS))
            return
        val altarNbt = nbt.getCompoundTag(TAG_ALTAR_POSITIONS)
        val spirits = ModMappers.spirits.getValues()
        val positions = HashMap<Int, MutableSet<BlockPos>>()
        for (spirit in spirits) {
            val spiritName = spirit.registryName!!.toString()
            if (!altarNbt.hasKey(spiritName))
                return
            val posArray = altarNbt.getIntArray(spiritName)
            val altarLocations = UtilBlockPos.deserialize(posArray)
            positions[spirit.id] = HashSet(altarLocations)
        }
        instance.allPositions = positions
    }

    companion object {
        private val TAG_ALTAR_POSITIONS = "AltarPositions"
    }
}
